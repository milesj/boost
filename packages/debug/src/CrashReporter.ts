/* eslint-disable no-magic-numbers */

import fs from 'fs';
import os from 'os';
import path from 'path';
import execa from 'execa';
import glob from 'fast-glob';
import { PortablePath, FilePath, requireModule, toArray, PackageStructure } from '@boost/common';
import { debug } from './constants';

function run(command: string, args: string[]): string {
  return String(execa.sync(command, args, { preferLocal: true }).stdout);
}

function resolveHome(filePath: FilePath): string {
  return filePath.replace(process.env.HOME!, '~');
}

function extractVersion(value: string): string {
  const match = value.match(/\d+\.\d+\.\d+([.-a-z0-9])?/u);

  return match ? match[0] : '';
}

export default class CrashReporter {
  contents: string = '';

  /**
   * Add a label with a value, or multiple values, to the last added section.
   */
  add(label: string, ...messages: (string | number | PortablePath)[]): this {
    this.contents += `${label}:\n`;
    this.contents += `  ${messages.map(String).join(' - ')}\n`;

    return this;
  }

  /**
   * Start a new section with a title.
   */
  addSection(title: string): this {
    this.contents += `\n${title.toUpperCase()}\n`;
    this.contents += `${'='.repeat(title.length)}\n\n`;

    debug('Reporting crash with %s', title);

    return this;
  }

  /**
   * Report Node.js related binary versions and paths.
   */
  reportBinaries(): this {
    this.addSection('Binaries');

    const bins = {
      node: 'Node',
      npm: 'NPM',
      yarn: 'Yarn',
    };

    Object.keys(bins).forEach((bin) => {
      try {
        this.add(
          bins[bin as keyof typeof bins],
          extractVersion(run(bin, ['--version'])),
          resolveHome(run('which', [bin])),
        );
      } catch {
        // Ignore
      }
    });

    return this;
  }

  /**
   * Report all environment variables.
   */
  reportEnvVars(): this {
    this.addSection('Environment');

    const keys = Object.keys(process.env).sort();

    keys.forEach((key) => {
      this.add(key, process.env[key]!);
    });

    return this;
  }

  /**
   * Report common programming language versions and paths
   */
  reportLanguages(): this {
    this.addSection('Languages');

    const languages = {
      bash: 'Bash',
      go: 'Go',
      javac: 'Java',
      perl: 'Perl',
      php: 'PHP',
      python: 'Python',
      ruby: 'Ruby',
      rustup: 'Rust',
    };

    // When running on OSX and Java is not installed,
    // OSX will interrupt the process with a prompt to install Java.
    // This is super annoying, so let's not disrupt consumers.
    // istanbul ignore next
    if (os.platform() === 'darwin') {
      delete languages.javac;
    }

    Object.keys(languages).forEach((bin) => {
      let version;

      try {
        version = extractVersion(run(bin, ['--version']));

        if (!version) {
          version = extractVersion(run(bin, ['version']));
        }
      } catch {
        // Ignore
      }

      if (version) {
        this.add(
          languages[bin as keyof typeof languages],
          version,
          resolveHome(run('which', [bin])),
        );
      }
    });

    return this;
  }

  /**
   * Report information about the current `process`.
   */
  reportProcess(): this {
    this.addSection('Process');
    this.add('ID', process.pid);
    this.add('Title', process.title);
    this.add('Timestamp', new Date().toISOString());
    this.add('CWD', process.cwd());
    this.add('ARGV', process.argv.map((v) => `- ${v}`).join('\n  '));

    return this;
  }

  /**
   * Report the stack trace for a defined `Error`.
   */
  reportStackTrace(error: Error): this {
    this.addSection('Stack Trace');
    this.contents += error.stack;
    this.contents += '\n';

    return this;
  }

  /**
   * Report information about the platform and operating system.
   */
  reportSystem(): this {
    this.addSection('System');
    this.add('OS', os.platform());
    this.add('Architecture', os.arch());
    this.add('CPUs', os.cpus().length);
    this.add('Uptime (sec)', os.uptime());
    this.add(
      'Memory usage',
      `${Math.round((process.memoryUsage().heapUsed / 1024 / 1024) * 100) / 100} MB`,
    );

    // istanbul ignore next
    if (process.platform !== 'win32') {
      this.add('Group ID', process.getgid());
      this.add('User ID', process.getuid());
    }

    return this;
  }

  /**
   * Report NPM package versions for all that match the defined pattern.
   * Only searches in the root node modules folder.
   */
  reportPackageVersions(patterns: string | string[], title: string = 'Packages'): this {
    this.addSection(title);

    const map = new Map<string, string>();

    glob
      .sync(
        toArray(patterns).map((pattern) => path.join('./node_modules', pattern)),
        {
          absolute: true,
          onlyDirectories: true,
          onlyFiles: false,
        },
      )
      .forEach((pkgPath) => {
        const pkg = requireModule<PackageStructure>(path.join(pkgPath, 'package.json'));

        map.set(pkg.name, pkg.version);
      });

    map.forEach((version, name) => {
      this.add(name, version);
    });

    return this;
  }

  /**
   * Write the reported content to the defined file path.
   */
  write(filePath: PortablePath): this {
    fs.writeFileSync(String(filePath), this.contents.trim(), 'utf8');

    return this;
  }
}
