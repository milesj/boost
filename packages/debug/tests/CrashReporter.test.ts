import fs from 'fs';
import execa from 'execa';
import CrashReporter from '../src/CrashReporter';

jest.mock('execa');

describe('CrashReporter', () => {
  let reporter: CrashReporter;
  let writeSpy: jest.SpyInstance;

  beforeEach(() => {
    (execa.sync as jest.Mock).mockImplementation((command, args = []) => {
      // Test the fallback to `version`
      if (command === 'php' && args[0] === '--version') {
        return {
          stderr: '',
          stdout: '',
        };
      }

      return {
        stderr: '',
        stdout: command === 'which' ? `/${args.join(' ')}` : '0.0.0',
      };
    });

    reporter = new CrashReporter();
    writeSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation(jest.fn());
  });

  afterEach(() => {
    writeSpy.mockRestore();
  });

  it('adds content and sections', () => {
    reporter.add('Label', 'Value');

    expect(reporter.contents).toMatchSnapshot();

    reporter.addSection('Section 1');
    reporter.add('Foo', 'Value');
    reporter.add('Bar', 'Value 1', 'Value 2');

    expect(reporter.contents).toMatchSnapshot();

    reporter.addSection('Section 2');
    reporter.add('Baz', 123);

    expect(reporter.contents).toMatchSnapshot();
  });

  it('writes content to the defined file path', () => {
    reporter.addSection('Section');
    reporter.add('Label', 'Value');
    reporter.write('./some.log');

    expect(writeSpy).toHaveBeenCalledWith(
      './some.log',
      `SECTION
=======

Label:
  Value`,
      'utf8',
    );
  });

  it('reports binaries', () => {
    reporter.reportBinaries();

    expect(reporter.contents).toMatchSnapshot();
  });

  it('reports environment variables', () => {
    const oldVars = process.env;

    process.env = {
      FOO: 'foo',
      BAR: 'bar',
      BAZ: 'baz',
    };

    reporter.reportEnvVars();

    expect(reporter.contents).toMatchSnapshot();

    process.env = oldVars;
  });

  it('reports languages', () => {
    reporter.reportLanguages();

    expect(reporter.contents).toContain('Ruby');
    expect(reporter.contents).toContain('0.0.0 - /php');
  });

  it('reports current process', () => {
    reporter.reportProcess();

    expect(reporter.contents).toContain('ID');
    expect(reporter.contents).toContain('CWD');
  });

  it('reports a stack trace', () => {
    reporter.reportStackTrace(new Error('Oops'));

    expect(reporter.contents).toContain('Error: Oops');
    expect(reporter.contents).toContain('at Object');
  });

  it('reports current system', () => {
    reporter.reportSystem();

    expect(reporter.contents).toContain('OS');
    expect(reporter.contents).toContain('CPUs');
  });

  it('reports package versions', () => {
    reporter.reportPackageVersions('path-*');

    expect(reporter.contents).toMatchSnapshot();
  });

  it('reports package versions (scoped)', () => {
    reporter.reportPackageVersions('@beemo/*', 'Beemo');

    expect(reporter.contents).toMatchSnapshot();
  });
});
