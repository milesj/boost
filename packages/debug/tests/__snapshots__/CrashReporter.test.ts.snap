// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrashReporter adds content and sections 1`] = `
"Label:
  Value
"
`;

exports[`CrashReporter adds content and sections 2`] = `
"Label:
  Value

SECTION 1
=========

Foo:
  Value
Bar:
  Value 1 - Value 2
"
`;

exports[`CrashReporter adds content and sections 3`] = `
"Label:
  Value

SECTION 1
=========

Foo:
  Value
Bar:
  Value 1 - Value 2

SECTION 2
=========

Baz:
  123
"
`;

exports[`CrashReporter reports binaries 1`] = `
"
BINARIES
========

Node:
  0.0.0 - /node
npm:
  0.0.0 - /npm
Yarn:
  0.0.0 - /yarn
"
`;

exports[`CrashReporter reports environment variables 1`] = `
"
ENVIRONMENT
===========

BAR:
  bar
BAZ:
  baz
FOO:
  foo
"
`;

exports[`CrashReporter reports package versions (scoped) 1`] = `
"
MOON
====

@moonrepo/dev:
  2.0.0
"
`;

exports[`CrashReporter reports package versions 1`] = `
"
PACKAGES
========

path-exists:
  4.0.0
path-is-absolute:
  1.0.1
"
`;
